### ISSUE TITLE
`ttnn.tanh` accuracy could be improved

### ISSUE THREAD

**Issue by sgligorijevicTT** [tenstorrent/tt-metal#25325]
### Describe the bug

`ttnn.tanh` specifically with small values seems to have a high error as measured by ULP difference.
It seems to be a known bug ( #23562 ) but I want to have a separate issue for tracking on Forge side.
This is a blocker for bringup of one (non priority model), VisionTextDualEncoder from tt-xla. This model has multiple outputs, and only the outputs that include tanh in their computation have bad accuracy. Furthermore, using tooling to bypass tanh on device makes the model pass.
As an interesting extra datapoint, another model(albert_xlarge) fails when using tanh, but we can detect the "gelu_new" block (which contains tanh) and replace the entire block with ttnn.gelu to make that model pass.

### Steps to reproduce the issue

1. Run the following code
```
import torch
import ttnn
import numpy as np

def spacing(x: torch.Tensor):
    """Calculate spacing (ULP size) for bfloat16 and float32 inputs."""
    # If input is bfloat16, use abs(x) * 2^-7 for normal numbers
    if x.dtype == torch.bfloat16:
        ulp = torch.abs(x).to(torch.float32) * 2**-7
        # For zero, use smallest positive normal bfloat16 (2^-126)
        ulp = torch.where(x == 0, torch.tensor(2**-126, dtype=torch.float32), ulp)
        return ulp
    else:
        # For float32, use nextafter
        direction = torch.where(x >= 0, torch.inf, -torch.inf)
        next_val = torch.nextafter(x, direction)
        return torch.abs(next_val - x)

def ulp_difference(a, b):
    """Calculate ULP (Units in the Last Place) difference between two tensors for bfloat16 and float32."""
    # Ensure inputs are torch tensors
    if not torch.is_tensor(a):
        a = torch.tensor(a)
    if not torch.is_tensor(b):
        b = torch.tensor(b)
    # Move to CPU and ensure same device/dtype
    a = a.detach().cpu()
    b = b.detach().cpu()
    # Handle special cases (inf, nan)
    mask_valid = torch.isfinite(a) & torch.isfinite(b)
    if not torch.any(mask_valid):
        return torch.tensor([float('inf')])
    # Use the larger absolute value for spacing calculation
    max_abs = torch.maximum(torch.abs(a), torch.abs(b))
    # Calculate spacing (distance to next representable float)
    ulp_spacing = spacing(max_abs)
    # Calculate ULP difference
    abs_diff = torch.abs(a - b)
    ulp_diff = abs_diff / ulp_spacing
    return ulp_diff[mask_valid]

RANGE = 3

def test_ttnn_gelu():
    x_values = torch.linspace(-RANGE, RANGE, 2048, dtype=torch.bfloat16).reshape(-1, 32, 32)
    
    with ttnn.manage_device(0) as dev:
        # Convert to ttnn tensor with BFLOAT16 and tile layout
        ttnn_input = ttnn.from_torch(
            x_values,
            dtype=ttnn.bfloat16,
            layout=ttnn.TILE_LAYOUT,
            device=dev
        )
    
        ttnn_result = ttnn.gelu(ttnn_input)
        ttnn_output = ttnn.to_torch(ttnn_result)
        torch_reference = torch.nn.functional.gelu(x_values)
        
        print("\n" + "="*60)
        print("COMPARISON RESULTS")
        print("="*60)
        
        # Calculate various error metrics
        abs_diff = torch.abs(ttnn_output - torch_reference)
        rel_diff = abs_diff.float() / (torch.abs(torch_reference.float()) + 1e-9)
        
        print(f"Max absolute difference: {abs_diff.max():.6e}")
        print(f"Mean absolute difference: {abs_diff.mean():.6e}")
        print(f"Max relative difference: {rel_diff.max():.6e}")
        print(f"Mean relative difference: {rel_diff.mean():.6e}")
        
        # Calculate ULP difference
        ulp_diffs = ulp_difference(ttnn_output, torch_reference)
        
        print(f"\nULP Statistics:")
        print(f"Max ULP difference: {torch.max(ulp_diffs):.2f}")
        print(f"Mean ULP difference: {torch.mean(ulp_diffs):.2f}")
        print(f"Median ULP difference: {torch.median(ulp_diffs):.2f}")
        
        N=30
        # Print some sample comparisons (equally spaced subset)
        total_values = len(x_values.flatten())
        indices = np.linspace(0, total_values - 1, N, dtype=int)
        
        print(f"\nSample comparisons ({N} equally spaced values):")
        print("Input\t\tTorch gelu\tTTNN gelu\tAbs Diff\tULP Diff\tOne ULP size")
        print("-" * 70)
        for i in indices:
            inp = x_values.flatten()[i].item()
            torch_val = torch_reference.flatten()[i].item()
            ttnn_val = ttnn_output.flatten()[i].item()
            abs_d = abs_diff.flatten()[i].item()
            ulp_d = ulp_diffs[i] if i < len(ulp_diffs) else 0
            torch_ulp = spacing(torch.tensor(torch_val, dtype=torch.bfloat16)).item()
            print(f"{inp:8.3f}\t{torch_val:10.6f}\t{ttnn_val:10.6f}\t{abs_d:.2e}\t{ulp_d:8.2f}\t{torch_ulp:.2e}")
        
        reasonable_ulp = torch.sum(ulp_diffs < 50) / len(ulp_diffs) * 100
        print(f"\nPercentage of values with ULP < 50: {reasonable_ulp:.1f}%")


if __name__ == "__main__":
    test_ttnn_gelu()

```

Note: The range being [-0.06, 0.06] for the test case is because in my model the values entering tanh seem to be around that size.

### Expected behavior

I am not sure what level of accuracy I need to make all models pass, but I do need better accuracy in general. 

### Please complete the following environment information

Ubuntu 22.04
Metal commit d5c259dd7bc8d0c9879a4389d50e65c321cc0c6f
N150 board

**Comment by sgligorijevicTT** [comment]
As there is only one model, not that popular and non-priority model being affected, I'm putting P2 priority for now.

**Comment by cglagovichTT** [comment]
This is also relevant for the Mochi video generation model.
In practice I've found that a usable workaround for correctness is to decompose tanh into `ttnn.sinh / ttnn.cosh`, though this can be a performance issue.

**Comment by nmauriceTT** [comment]
@cglagovichTT 
Thank you for the information. 
The new exp being merged ( #25946 ) should hopefully unblock the accuracy problem for tanh. An optimization round on exp is also on the menu to maintain performance as much as possible.

**Comment by cglagovichTT** [comment]
Thanks @nmauriceTT. Do you have issues I could be tracking?

**Comment by nmauriceTT** [comment]
@cglagovichTT I have opened an issue for the micro-optimization of the new exp: https://github.com/tenstorrent/tt-metal/issues/26988

**Comment by fahmedchanTT** [comment]
Bumping to P1:
- Useful for Forge
- Useful for Video
- Useful for DX

**Comment by nmauriceTT** [comment]
Hello @sgligorijevicTT @cglagovichTT ,

 I've merged PR #31685  that should have significantly increased the accuracy of tanh (especially in the range near 0) + increased execution speed.
